### ISSUE TITLE
Unsaved scripts don't prompt when closing editor

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot#55026]
### Godot version

098e3cf / 3.4

### System information

W10

### Issue description

When you modify a script, it gets the "modified" status, but it's ignored by the editor. When you quit, any script changes are lost. Only unsaved scenes are handled.

Resolving this would effectively resolve #36281 (after #54653)

### Steps to reproduce

1. Open any script in script editor
2. Modify it
3. Close Godot

**Comment by ator-dev** [comment]
Sorry, I had this open in a different tab and copied the wrong issue tag... oops

**Comment by ator-dev** [comment]
@KoBeWi Should a 'save script' dialogue appear for each unsaved script, or just a 'save all script changes' overall dialogue; or should this be an option toggled via the settings like for scenes?
Alternatively, should script changes just count as part of their scene/s, and not show any additional popups?

**Comment by ator-dev** [comment]
@Calinou I have an effectively complete solution working at this point, but I've run into a problem. Built-in scripts don't seem to get a 'modified' status (though they are recognised as unsaved when quitting), and furthermore aren't saved by the `save_all_scripts` method meaning that my code doesn't work for them. I can't work out how to _actually_ save all scripts, or to do it separately for built-in and external scripts (without saving scenes at the same time).
Do you have any advice for how I could resolve this, or will #54653 be adapted for 3.x (making both script types save in a similar way), meaning I should just pull request anyway?

**Comment by KoBeWi** [comment]
The PR should be made against master branch first.

**Comment by ator-dev** [comment]
Ah, ok - because the 3.4 version is included in this issue I assumed it was for 3.x, so made my changes on that version. Should I make separate pull requests for 3.x and 4.0? I don't really know how it's done on GitHub yet.

2 Dec 2021 17:33:05 Tomek ***@***.***>:

**Comment by KoBeWi** [comment]
> Should I make separate pull requests for 3.x and 4.0?

Depends on how much different they are. Sometimes the changes are the same for both branches, so the commit can be cherry-picked. But if they are too different, 3.x needs a separate PR.

**Comment by ator-dev** [comment]
That makes sense, thank you (;

2 Dec 2021 17:55:16 Tomek ***@***.***>:

**Comment by ator-dev** [comment]
There's one aspect of my changes (and my changes in 3.x) that I don't like; I had to add a variable `bool close_without_saving` to script_editor_plugin, as I couldn't find a better way of preventing scripts from being automatically saved on quit. I've added a TODO comment to this and might find a different workaround later. Apart from that, my fixes worked as expected in my tests and seem to fit in with the rest of the code workings.

**Comment by me2beats** [comment]
Is this still relevant?
I tested it on 3.5 beta — scripts are saved when closing the editor (there's no prompt tho)
I tried different editor options.

**Comment by me2beats** [comment]
I also tested it in 3.4 — scripts are saved
Win7

**Comment by KoBeWi** [comment]
> there's no prompt tho

That's a bug: #18010
There should be a prompt. Also did you test with built-in scripts?

**Comment by me2beats** [comment]
Well yeah built-in scripts are not saved when project reloading.
So is this only built-in scripts problem (script changes are lost)

**Comment by ator-dev** [comment]
> I had to add a variable `bool close_without_saving`

Just providing an update: I found my PR again recently, and have reworked it to work again and to remove poor practices such as the above unnecessary statefulness. I'm happy with how it works now, though shaders and built-in scripts would definitely benefit from special handling in future.

> built-in scripts are not saved when project reloading

Built-in scripts are saved when their scene is saved, which means there is some additional weirdness and inconsistency in the saving system. I've addressed this temporarily by displaying a warning with the list of unsaved built-in scripts (since their scenes will be saved without prompting), but it should be addressed fully whenever possible.
